deployment/WFCORE-3522_Deployment content validation.adoc

```diff
@@ -0,0 +1,72 @@
+= Title
+:author:            David Lloyd
+:email:             david.lloyd@redhat.com
+:toc:               left
+:icons:             font
+:keywords:          deployment
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+We need to move to child-first class and resource resolution in deployments.
+The first step must be validation of deployment contents and dependencies.
+This entails detecting deployment artifacts which have packages or resources which conflict with container packages or resources.
+
+Once this proposal is implemented, we can move to changing dependency order, which will come in a follow-up proposal.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFCORE-3522[WFCORE-3522]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/SWARM-1691[SWARM-1691]
+* https://issues.jboss.org/browse/WFLY-9828[WFLY-9828]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* mailto:mjurc@redhat.com[Michal Jurc]
+
+=== Affected Projects or Components
+
+* WildFly Core
+* All deployment-related subsystems
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+* The deployer must provide a way for subsystems to register packages and resources that are indicative of invalid deployment content.
+This likely entails introducing a multi-valued capability which is consumed by the deployment system.
+** The deployer must correlate this information in such a way that if the subsystem is disabled for a given deployment via ``exclude-subsystem`` in ``jboss-deployment-structure.xml``, the deployment can proceed without error.
+** There may be APIs that must always be forbidden, even when the ``exclude-subsystem`` flag is set.
+In these cases, the subsystem must have a way to indicate this situation.
+* The deployer must ignore invalid content.
+* The deployer must produce clear, user-friendly warnings about invalid content, ideally with a suggested remedy.
+
+=== Nice-to-Have Requirements
+
+* Implement an optional deployment mode where users could specify that they wish to reject deployments with invalid content.
+This could become the default setting in some future version.
```

Comment:
Is there any reasoning against default fast fail on deployment? I think it could be more user friendly (for starters).

Reply:
It would be a compatibility breakage to do so, unfortunately.

Reply:
One possibility would be to use an attribute to configure whether invalid content should be ignored or rejected.  The default shipped configuration could be our preferred behavior which is reject invalid content, but the transformers for older model versions could use a default value of "ignored" for compatibility.  @bstansberry are we allowed to do things like that?

Reply:
Answering a very old question I just saw. :)

No that would not be valid. Behavior should be consistent across servers using a profile, so if the standard config specifies something a server can't support that should result in a failure when the server joins the domain.

Our standard profiles need not work with legacy servers. Any time we introduce a new subsystem they don't. Users who want to run older version servers should adapt our profiles to suit their needs or introduce new ones tailored to those older servers. The EAP docs emphasize the latter approach.

It is certainly possible though to have the default value for a new attribute be different from what our standard configs specify, with the default resulting in previous behavior. That can be helpful for migration as existing configs still work as they did before.